.githooks/pre-push

```diff
@@ -1,5 +1,6 @@
 #!/usr/bin/env sh
 
+git stash -q --keep-index
```

Comment:
what if there are no items to stash?

Reply:
Do you mean that we should stash untracked files also? 
If yes, I think we need to do that.

Reply:
nope -- but yes we should stash untracked files too, e.g. there may be issues with import checks / running tests from newly added files. Nice catch 💯 

More so simply: if there's nothing to stash, no stash would be created, in which case we shouldn't try running `pop` at all (e.g. it might pop an older manual stash)

Reply:
The flag --keep-index cancels removal of the cached changes from the working directory (but they are still included in the stash). (Explained by [stackoverflow](https://stackoverflow.com/questions/69110050/in-git-pre-commit-hook-temporarily-remove-all-changes-that-are-not-about-to-be))

So, a stash would still be created and I tested it on my end. (That seems to be the case)

Reply:
Hmm to check we're on the same page, here's an example, I don't think the `keep-index` option is for this:

1. I'm working on XX branch, I run `git stash`
2. I switch to YY branch, I `git add .` *all* changes
3. `git commit -m "..."`
4. In the script, you're running `git stash -q --include-untracked --keep-index`. This does nothing since I've `git add .` everything. (That seems to be the case on my end at least as well)
5. `git stash pop -q` accidentally pops my old stash from XX branch.

Reply:
There would be things to stash as the staged files would be stashed. With --keep-index, we will stash the staged files but  not remove it. So, something will still be stashed.
Does this behaviour still occur when you run attempt to commit. It seems working on my end hmmm.

Reply:
Managed to get it to work, I needed to actually make some modifications (was testing with no changes at all). Thanks for the link!

> was testing with no changes at all

The XX branch stash is still popped here. Should we do something simple like naming the stash (with something very unique) and popping only that named stash to prevent this?

Reply:
Another issue seems to be file renames, in this case the rename itself is stashed (`--keep-index` dosen't seem to work literally). Is this occuring for you as well?

Reply:
> prevent this

I think could also help prevent other things we haven't accounted for in the discussion here. 🙂

Reply:
one downside/potential issue of naming might be if the script gets in a bad state (leftover stash from previous precommit run)

or we can try to make the name unique only within the same precommit run.. 💭

Reply:
@ang-zeyu 
I tried doing that by making the name unique but then I realised that you cannot drop/pop base on stash name. You can only drop or pop by index.

So, what I am doing will be checking the stash count before and after stashing to determine whether to pop.

Reply:
> cannot drop/pop base on stash name

Does this work? https://stackoverflow.com/questions/11269256/how-do-i-name-and-retrieve-a-git-stash-by-name

> So, what I am doing will be checking the stash count before and after stashing to determine whether to pop.

That works too!

Reply:
`git stash pop stash^{/my_stash_name}`
There are quite a lot of guide that states this but this does not actually works.

Reply:
> --keep-index

There's an issue here:
1. Change some line to trigger a lint fail (e.g. add trailing whitespace), stage it
2. Change it again, don't stage it
3. Run your precommit script
4. On stash pop, a merge conflict occurs

Would the alternative pointed out here work? https://stackoverflow.com/questions/20028507/git-stash-uncached-how-to-put-away-all-unstaged-changes/34681302#34681302

(make a temp commit without running pre commit, stash, soft reset, continue)

iirc (it's been a while since I used it, do double check 😅) [precommit tool](https://pre-commit.com/) dosen't have this issue too, could also reference how it does it.

Reply:
> merge conflict

presumably this occurs as the staged changes are also stashed.

Reply:
>You can only apply but you cannot pop.

That makes sense, per https://git-scm.com/docs/git-stash.

I think apply has to be followed by a `drop <stash>` (if this works too)

Reply:
> Change some line to trigger a lint fail (e.g. add trailing whitespace), stage it

![Screenshot 2022-10-03 205951](https://user-images.githubusercontent.com/3306138/193582868-13310f92-35e6-4e21-9310-8c826b827d70.png)

> Change it again, don't stage it

![Screenshot 2022-10-03 210020](https://user-images.githubusercontent.com/3306138/193582913-c3aa9421-dc42-48c4-a32f-6f85132471a4.png)

> On stash pop, a merge conflict occurs

![Screenshot 2022-10-03 210104](https://user-images.githubusercontent.com/3306138/193582948-9fef2a0f-dc15-4fc3-b79c-8f610763f89c.png)